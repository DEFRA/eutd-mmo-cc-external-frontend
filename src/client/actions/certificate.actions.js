import { isError, showFullPageError } from './index';

export const GET_SUMMARY_DOCUMENT = 'GET_SUMMARY_DOCUMENT';
export const SHOW_SUMMARY_DOCUMENT_ERRORS = 'SHOW_SUMMARY_DOCUMENT_ERRORS';
export const GET_SUMMARY_UNAUTHORISED = 'GET_SUMMARY_DOCUMENT_UNAUTHORISED';

export function showSummaryCertificateErrors(data) {
  return { type: SHOW_SUMMARY_DOCUMENT_ERRORS, payload: { validationErrors: data } };
}

export const getSummaryCertificate = (journey, documentNumber = undefined) => async (dispatch, getState, { orchestrationApi }) => {
  const config = (documentNumber)
    ? { headers: { documentnumber: documentNumber } }
    : {};

  await orchestrationApi.get(`/certificate/${journey}`, config)
    .then(
      res => {
        dispatch({ type: GET_SUMMARY_DOCUMENT, payload: res.data });
      },
      error => {
        if (isError(error)) {
          dispatch(showFullPageError());
        } else {
          dispatch(unauthorised());
        }
      });

    function unauthorised() { return { type: GET_SUMMARY_UNAUTHORISED }; }
};