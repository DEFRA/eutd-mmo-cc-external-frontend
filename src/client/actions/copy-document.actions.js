import { showFullPageError, apiCallFailed, isError, CLEAR_ERRORS } from './index';
export const CONFIRM_COPY_DOCUMENT            = 'confirm_copy_document';
export const UNAUTHORISED_COPY_DOCUMENT       = 'unauthorised_copy_document';
export const CLEAR_COPY_DOCUMENT              = 'clear_copy_document';
export const REMOVE_CONFIRM_COPY_CERTIFICATE  = 'remove_confirm_copy_certificate';
export const ADD_ERROR_COPY_CERTIFICATE       = 'add_error_copy_certificate';
export const REMOVE_ERROR_COPY_CERTIFICATE    = 'remove_error_copy_certificate';

export const submitCopyCertificate = (payload) => async (dispatch, _getState, {orchestrationApi}) => {
  const {documentNumber, copyDocumentAcknowledged, voidOriginal, excludeLandings, journey} = payload;
  const config = {headers: {documentnumber: documentNumber}};

  try {
    let ipAddress;
    if (voidOriginal) {
      const ip = await orchestrationApi.get('/client-ip');
      ipAddress = ip.data;
    }

    return await orchestrationApi.post('/confirm-copy-certificate', { copyDocumentAcknowledged, voidOriginal, excludeLandings, journey, ipAddress }, config)
    .then(res => {
      dispatch({ type: CONFIRM_COPY_DOCUMENT, payload: {
        documentNumber: res.data.documentNumber,
        copyDocumentAcknowledged: res.data.copyDocumentAcknowledged,
        voidDocumentConfirm: res.data.voidOriginal
      }});
      return res.data.newDocumentNumber;
    });
  } catch (e) {
    if (isError(e)) dispatch(showFullPageError());
    if (e.response && e.response.status === 403) {
      dispatch({ type: UNAUTHORISED_COPY_DOCUMENT }); return;
    } else {
      dispatch(apiCallFailed(e.response));
    }
  }
};

export const checkCopyCertificate = (documentNumber) => async (dispatch, _getState, {orchestrationApi}) => {
  const config = {
    headers: {
      documentnumber: documentNumber
    }
  };

  try {
    const res = await orchestrationApi.get('/check-copy-certificate', config);
    const { canCopy } = res.data;
    if (!canCopy) {
      dispatch(unauthorised());
    }
  } catch(e) {
    if (isError(e)) dispatch(showFullPageError());
    if (e.response && e.response.status === 403) {
      dispatch(unauthorised());
      throw new Error(`Unauthorised access ${e.response.status}`);
    } else {
      dispatch(apiCallFailed(e.response));
      throw new Error(`An error has occurred ${e.response.status} (${e.response.statusText})`);
    }
  }

  function unauthorised() { return { type: UNAUTHORISED_COPY_DOCUMENT }; }
};

export const addCopyDocument = (data) => (dispatch) => dispatch({ type: CONFIRM_COPY_DOCUMENT, payload: data });
export const clearErrors = () => (dispatch) =>  dispatch({ type: CLEAR_ERRORS });
export const clearCopyDocument = () => (dispatch) => dispatch({ type: CLEAR_COPY_DOCUMENT });
export const unauthorisedCopyDocument = () => (dispatch) => dispatch({ type: UNAUTHORISED_COPY_DOCUMENT });