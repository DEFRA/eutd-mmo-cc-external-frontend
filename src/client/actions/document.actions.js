// actions
import {isError, showFullPageError} from './index';

export const GOT_ALL_DOCUMENTS          = 'GOT_ALL_DOCUMENTS';
export const GET_ALL_DOCUMENTS_FAILED   = 'GET_ALL_DOCUMENTS_FAILED';
export const GET_COMPLETED_DOCUMENT = 'GET_COMPLETED_DOCUMENT';
export const GET_PENDING_DOCUMENT = 'GET_PENDING_DOCUMENT';


// action creators
export const getAllDocuments = (type, month, year) => async (dispatch, getState, { orchestrationApi }) => {
  try {
    const res = await orchestrationApi.get(`/documents/${year}/${month}?type=${type}`);
    return dispatch({ type: GOT_ALL_DOCUMENTS, payload: res.data });

  } catch(e) {
    if( isError(e) ) dispatch(showFullPageError());
    //throw new Error(e.response);
  }
};

export const getCompletedDocument = (documentNumber) => async (dispatch, getState, { orchestrationApi }) => {
  try {
    const res = await orchestrationApi.get(`/document/${documentNumber}`);
    if (res.data && res.data.documentStatus === 'COMPLETE') {
      return dispatch({ type: GET_COMPLETED_DOCUMENT, payload: res.data });
    }
    return dispatch({ type: GET_COMPLETED_DOCUMENT, payload: null });
  }
  catch (e) {
    return dispatch({ type: GET_COMPLETED_DOCUMENT, payload: null });
  }
};

export const getPendingDocument = (documentNumber) => async (dispatch, getState, { orchestrationApi }) => {
  try {
    const res = await orchestrationApi.get(`/document/${documentNumber}`);
    if (res.data && res.data.documentStatus === 'PENDING') {
      return dispatch({ type: GET_PENDING_DOCUMENT, payload: res.data });
    }
    return dispatch({ type: GET_PENDING_DOCUMENT, payload: null });
  }
  catch (e) {
    return dispatch({ type: GET_PENDING_DOCUMENT, payload: null });
  }
};