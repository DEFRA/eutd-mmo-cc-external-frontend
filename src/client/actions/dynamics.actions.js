const {parseContactsResponse, parseAddressResponse, parseAccountsResponse} = require('../../helpers/dynamix');

export const dynamicsActionTypes = {
  USER_DETAILS_REQUESTED: 'user-details/requested',
  USER_DETAILS_LOADED: 'user-details/loaded',
  USER_DETAILS_FAILED: 'user-details/failed',
  ADDRESS_DETAILS_REQUESTED: 'address-details/requested',
  ADDRESS_DETAILS_LOADED: 'address-details/loaded',
  ADDRESS_DETAILS_FAILED: 'address-details/failed',
  ACCOUNT_DETAILS_REQUESTED: 'account-details/requested',
  ACCOUNT_DETAILS_LOADED: 'account-details/loaded',
  ACCOUNT_DETAILS_FAILED: 'account-details/failed',
  EXPORTER_NAME_PRE_LOADED: 'export-certificate/exporter-name-preloaded',
  EXPORTER_ADDRESS_PRE_LOADED: 'export-certificate/exporter-address-preloaded',
  EXPORTER_ADDRESS_PRE_LOAD_FAILED: 'export-certificate/exporter-address-preload-failed',
  EXPORTER_COMPANY_NAME_PRE_LOADED: 'export-certificate/exporter-company-name-preloaded'
};


export const fetchUserDetailsFromDynamics = () => async (dispatch, getState, { dynamixApi } ) => {
  const url = '/user-details';
  try {
    const res = await dynamixApi.get(url);
    let userdetails = parseContactsResponse(res);
    dispatch(success(userdetails));
    if (userdetails && userdetails.length > 0) {
      dispatch(exporterNamePreloaded(userdetails[0].firstName, userdetails[0].lastName, userdetails[0].contactId));
    }
  } catch(error) {
    dispatch(failed('Failed server call to get user details'));
  }

  function success(userdetails) { return { type: dynamicsActionTypes.USER_DETAILS_LOADED, userdetails }; }
  function failed(error) { return { type: dynamicsActionTypes.USER_DETAILS_FAILED, error }; }
  function exporterNamePreloaded(firstName, lastName, contactId) { return { type: dynamicsActionTypes.EXPORTER_NAME_PRE_LOADED, firstName, lastName, contactId }; }
};

export const fetchAccountDetailsFromDynamics = () => async (dispatch, getState, { dynamixApi } ) => {
  function success(accounts) { return { type: dynamicsActionTypes.ACCOUNT_DETAILS_LOADED, accounts }; }
  function failed(error) { return { type: dynamicsActionTypes.ACCOUNT_DETAILS_FAILED, error }; }
  function exporterCompanyNamePreloaded(name, accountId) { return { type: dynamicsActionTypes.EXPORTER_COMPANY_NAME_PRE_LOADED, name, accountId }; }
  const url = '/accounts';

  try {
    const res = await dynamixApi.get(url);
    let accounts = parseAccountsResponse(res);
    dispatch(success(accounts));
    if (accounts && accounts.length > 0) {
      dispatch(exporterCompanyNamePreloaded(accounts[0].name, accounts[0].accountId));
    }
  } catch(error) {
    dispatch(failed('Failed server call to get account details'));
  }
};

export const fetchAddressDetailsFromDynamics = () => (dispatch, getState, { dynamixApi } ) => {
  const url = '/addresses';
  return dynamixApi.get(url)
    .then(
      res => {
        let addresses = parseAddressResponse(res);
        dispatch(success(addresses));
        if (addresses && addresses.length > 0) {
          dispatch(exporterAddressPreloaded(addresses[0]));
        }
      }
    )
    .catch((error) => {
      console.error(error);
      dispatch(failed('Failed server call to get address details'));
      dispatch(exporterAddressPreloadFailed());
    });

  function success(addresses) { return { type: dynamicsActionTypes.ADDRESS_DETAILS_LOADED, addresses }; }
  function failed(error) { return { type: dynamicsActionTypes.ADDRESS_DETAILS_FAILED, error }; }
  function exporterAddressPreloaded(address) { return { type: dynamicsActionTypes.EXPORTER_ADDRESS_PRE_LOADED, address }; }
  function exporterAddressPreloadFailed() { return { type: dynamicsActionTypes.EXPORTER_ADDRESS_PRE_LOAD_FAILED }; }
};

export const getAddressArrayFromDynamicsResponse = function(dynamicsResponse) {

  let addressArray = [];
  let addrOne = '';

  if (dynamicsResponse.address.premises && dynamicsResponse.address.premises !== null && dynamicsResponse.address.premises.length > 0) {
    addrOne += dynamicsResponse.address.premises;
    addrOne += ', ';
  }

  if (dynamicsResponse.address.subbuildingname && dynamicsResponse.address.subbuildingname !== null && dynamicsResponse.address.subbuildingname.length > 0) {
    addrOne += dynamicsResponse.address.subbuildingname;
    addrOne += ', ';
  }

  if (dynamicsResponse.address.buildingname && dynamicsResponse.address.buildingname !== null && dynamicsResponse.address.buildingname.length > 0) {
    addrOne += dynamicsResponse.address.buildingname;
    addrOne += ', ';
  }

  if (dynamicsResponse.address.street && dynamicsResponse.address.street !== null && dynamicsResponse.address.street.length > 0) {
    var partsOfStr = dynamicsResponse.address.street.split(',');
    if (partsOfStr.length > 1) {
      addrOne += partsOfStr[0].trim();
      addressArray.push(addrOne);
      addressArray.push(dynamicsResponse.address.street.slice(partsOfStr[0].length + 1).trim());
    } else {
      addrOne += dynamicsResponse.address.street;
      addressArray.push(addrOne);
    }
  } else {
    addressArray.push(addrOne);
  }

  if (dynamicsResponse.address.locality && dynamicsResponse.address.locality != null && dynamicsResponse.address.locality.length > 0) {
    addressArray.push(dynamicsResponse.address.locality);
  }
  if (dynamicsResponse.address.dependantLocality && dynamicsResponse.address.dependantLocality != null && dynamicsResponse.address.dependantLocality.length > 0) {
    addressArray.push(dynamicsResponse.address.dependantLocality);
  }
  if (dynamicsResponse.address.towntext && dynamicsResponse.address.towntext != null && dynamicsResponse.address.towntext.length > 0) {
    addressArray.push(dynamicsResponse.address.towntext);
  }
  if (dynamicsResponse.address.county && dynamicsResponse.address.county != null && dynamicsResponse.address.county.length > 0) {
    addressArray.push(dynamicsResponse.address.county);
  }

  return addressArray;
};

export const getAddressModelFromDynamicsResponse = function(dynamicsResponse) {

  let addressArray = getAddressArrayFromDynamicsResponse(dynamicsResponse);
  let model = {};
  model.addressOne = addressArray[0];

  if (dynamicsResponse.address.postcode && dynamicsResponse.address.postcode != null && dynamicsResponse.address.postcode.length > 0) {
    model.postcode = dynamicsResponse.address.postcode;
  } else if (dynamicsResponse.address.internationalpostalcode && dynamicsResponse.address.internationalpostalcode != null
    && dynamicsResponse.address.internationalpostalcode.length > 0) {
    model.postcode = dynamicsResponse.address.internationalpostalcode;
  }

  model._dynamicsAddress = dynamicsResponse.address._dynamicsAddress;

  let arrLength = addressArray.length;
  if (arrLength > 1) {
    model.townCity = addressArray[arrLength - 1];
    arrLength = arrLength - 1;
  }

  let addrTwo = '';
  if (arrLength > 1) {
    for (let idx = 1; idx < arrLength; idx++) {
      addrTwo += addressArray[idx];
      if (idx < arrLength - 1) {
        addrTwo += ', ';
      }
    }
  }
  model.addressTwo = addrTwo;
  model.buildingNumber = dynamicsResponse.address.premises;
  model.subBuildingName = dynamicsResponse.address.subbuildingname;
  model.buildingName = dynamicsResponse.address.buildingname;
  model.streetName = dynamicsResponse.address.street;
  model.townCity = dynamicsResponse.address.towntext || dynamicsResponse.address.locality;
  model.county = dynamicsResponse.address.county;
  model.country = dynamicsResponse.address.country;

  return model;
};

export const dynamicsActions = {
  fetchUserDetailsFromDynamics,
  fetchAccountDetailsFromDynamics,
  fetchAddressDetailsFromDynamics
};

