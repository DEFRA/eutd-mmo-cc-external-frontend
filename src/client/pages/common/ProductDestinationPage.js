import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import _ from 'lodash';

import { Main, BackLink, Header, GridRow, GridCol } from 'govuk-react';

import {
  addSelectedExportCountry,
  saveExportCountry,
  getExportCountry,
  getAllCountries,
  clearExportCountry
} from '../../actions';

import Form from '../../components/elements/Form';
import HelpLink from '../../components/HelpLink';
import PageTitle from '../../components/PageTitle';
import SaveAsDraftButton from '../../components/SaveAsDraftButton';
import ContinueButton from '../../components/elements/ContinueButton';
import ExportDestination from '../../components/ExportDestination';
import ErrorIsland from '../../components/elements/ErrorIsland';
import BackToProgressLink from '../../components/BackToProgressLink';
import { onHandleErrorClick } from '../utils';
import PageTemplateWrapper from '../../components/PageTemplateWrapper';
import { withTranslation } from 'react-i18next';

class ProductDestinationPage extends Component {
  constructor(props) {
    super(props);

    this.state = {
      exportDestination: '',
    };
  }

  doSubmit = async (isSaveAsDraft) => {
    const { path, nextUri } = this.props.route;
    const documentNumber =
      this.props.match && this.props.match.params
        ? this.props.match.params['documentNumber']
        : '';

    await this.props.saveExportCountry(
      path,
      nextUri,
      documentNumber,
      isSaveAsDraft
    );
  };

  onDestinationChange = (countryName) => {
    const exportLocation = _.cloneDeep(
      this.props.exportLocation || {}
    );

    exportLocation.exportedTo = countryName;

    this.props.addSelectedExportCountry(exportLocation);
  };

  onSaveAsDraft = async (e) => {
    e.preventDefault();
    const { saveAsDraftUri } = this.props.route;

    await this.doSubmit(true);
    this.props.history.push(saveAsDraftUri);
  };

  onSubmit = async (e) => {
    e.preventDefault();
    const { nextUri } = this.props.route;
    const documentNumber =
      this.props.match && this.props.match.params
        ? this.props.match.params['documentNumber']
        : '';

    await this.doSubmit(false);
    this.props.history.push(nextUri.replace(':documentNumber', documentNumber));
  };

  componentWillUnmount() {
    this.props.clearExportCountry();
  }

  componentDidMount = async () => {
    window.scrollTo(0, 0);

    this.setState({ jsEnabled: true });

    const documentNumber = this.props.match.params['documentNumber'];

    try {
      await this.props.getExportCountry(documentNumber);
      await this.props.getAllCountries();
    } catch (e) {
      console.error(e);
    }

    if (this.props.exportLocation.unauthorised === true) {
      this.props.history.push('/forbidden');
    }
  };

  render = () => {
    const { route, match, countries, exportLocation, errors, t } = this.props;
    const { title, header, path, nextUri, previousUri, journey, progressUri } = route;
    const { documentNumber } = match.params;

    const saveAsDraftFormActionUrl =
      '/orchestration/api/v1/exportDestination/saveAsDraftLink';
    const backLink = previousUri.replace(':documentNumber', documentNumber);

    const exportedTo =
      exportLocation && exportLocation.exportedTo
        ? exportLocation.exportedTo
        : null;
    if (!exportLocation.loaded) return null;

    const exportDestinationError =
      errors && Object.prototype.hasOwnProperty.call(errors, 'exportDestinationError')
        ? this.props.errors['exportDestinationError']
        : null;

    return (
      <Main>
        <PageTitle title={`${t(header)} - ${t(title)}`} />
        {!_.isEmpty(errors) && errors.errors && errors.errors.length > 0 && (
          <ErrorIsland
            errors={errors.errors.map((err) => ({
              message: t(err.text),
              key: err.targetName,
            }))}
            onHandleErrorClick={onHandleErrorClick}
          />
        )}
        <BackLink href={backLink}>
          {t('commonBackLinkBackButtonLabel')}
        </BackLink>
        <Header>{t(header)}</Header>
        <Form
          action="/orchestration/api/v1/exportDestination"
          currentUrl={path}
          nextUrl={nextUri}
          onSubmit={this.onSubmit}
        >
          <input type="hidden" name="journey" value={journey} />
          <input type="hidden" name="currentUri" value={path} />
          <GridRow>
            <GridCol columnTwoThirds>
              <ExportDestination
                name="exportDestination"
                exportDestination={exportedTo}
                countries={countries}
                onChange={this.onDestinationChange}
                error={exportDestinationError}
              />
            </GridCol>
          </GridRow>
          <GridRow>
            <SaveAsDraftButton
              formactionUrl={saveAsDraftFormActionUrl}
              onClick={this.onSaveAsDraft}
            />
            <ContinueButton type="submit" id="continue">
              {t('commonContinueButtonSaveAndContinueButton')}
            </ContinueButton>
          </GridRow>
        </Form>
        <BackToProgressLink
          progressUri={progressUri}
          documentNumber={documentNumber}
        />
        <HelpLink journey={journey} />
      </Main>
    );
  };
}

function mapStateToProps(state) {
  return {
    exportLocation: state.exportLocation,
    countries: state.global.allCountries,
    errors: state.errors,
  };
}

function loadData(store) {
  return store
    .dispatch(getExportCountry(this.documentNumber))
    .then(() => store.dispatch(getAllCountries()));
}

export const component = withRouter(
  connect(mapStateToProps, {
    addSelectedExportCountry,
    saveExportCountry,
    getExportCountry,
    getAllCountries,
    clearExportCountry
  })(withTranslation() (ProductDestinationPage))
);

export default {
  loadData,
  component: PageTemplateWrapper(component),
};
