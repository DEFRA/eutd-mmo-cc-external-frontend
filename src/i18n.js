import i18n from 'i18next';
import  {initReactI18next}  from 'react-i18next';

import en_UK from './locales/uk/en.json';
import cy_UK from './locales/uk/cy.json';

const resources = {
  en_UK: {
    translation: en_UK,
  },
  cy_UK: {
    translation: cy_UK,
  }
};

i18n
  .use(initReactI18next).init({
    resources,
    fallbackLng: 'en_UK',
    keySeparator: '.',
    interpolation: {
      escapeValue: false
    }
});

export default i18n;
