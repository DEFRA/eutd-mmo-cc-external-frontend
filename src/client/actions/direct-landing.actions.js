import { beginApiCall, apiCallFailed, isForbidden, showFullPageError, CLEAR_ERRORS, isSilverLineError } from './index';
import { exportPayloadActionTypes } from '../actions';


export const GET_DIRECT_LANDING_PRODUCTS = 'get_direct_landing_products';
export const CLEAR_DIRECT_LANDING = 'clear_direct_landing';
export const DIRECT_LANDING_PRODUCTS_FAILED = 'direct_landing_products_failed';
export const POST_DIRECT_LANDING_PRODUCTS_UNAUTHORISED = 'post_direct_landing_products_unauthorised';
export const GET_DIRECT_LANDING_PRODUCTS_UNAUTHORISED = 'get_direct_landing_products_unauthorised';
export const POST_DIRECT_LANDING_PRODUCTS_WAF_ERROR = 'post_direct_landing_prodcts_waf_error';

export const getDirectLandings =
  (documentNumber) =>
  async (dispatch, _getState, { orchestrationApi }) => {
    const config = { headers: { documentnumber: documentNumber } };

    try {
      dispatch(beginApiCall());

      const res = await orchestrationApi.get(
        '/export-certificates/export-payload/direct-landings',
        config
      );

      await dispatch({ type: GET_DIRECT_LANDING_PRODUCTS, payload: res.data });
    } catch (e) {
      if (isForbidden(e)) {
        dispatch({ type: GET_DIRECT_LANDING_PRODUCTS_UNAUTHORISED });
      } else {
        dispatch(apiCallFailed(e.response));
        throw new Error(
          `An error has occurred ${e.response.status} (${e.response.statusText})`
        );
      }
    }
  };

export const clearDirectLanding = () => ({ type: CLEAR_DIRECT_LANDING });

export const upsertDirectLanding = (landing, documentNumber = undefined) => (dispatch, getState, { orchestrationApi } ) => {
  const config = (documentNumber)
    ? {
        headers: {
          documentnumber: documentNumber
        }
    }
    : {};

  return orchestrationApi.post('/export-certificates/direct-landing/validate', landing, config)
    .then(
      res => {
        dispatch(landingUpserted(res.data));
        dispatch(clearErrors());
      },
      error => {
        if (error.response.status === 400) {
          dispatch(landingUpserted(error.response.data));
          throw new Error();
        } else if (isSilverLineError(error)) {
          dispatch({ type: POST_DIRECT_LANDING_PRODUCTS_WAF_ERROR, supportID: error.response.data.match(/\d+/)[0] });
        } else if (isForbidden(error)) {
          dispatch({ type: POST_DIRECT_LANDING_PRODUCTS_UNAUTHORISED });
        } else {
          dispatch(showFullPageError());
        }
      });

  function landingUpserted(exportPayload) { return { type: exportPayloadActionTypes.LANDING_UPSERTED, exportPayload, errors: exportPayload.errors }; }
  function clearErrors() { return { type: CLEAR_ERRORS }; }
};
