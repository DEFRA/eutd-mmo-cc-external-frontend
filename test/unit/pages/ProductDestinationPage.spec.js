import { mount } from 'enzyme';
import * as React from 'react';
import { component as ProductDestinationPage } from '../../../src/client/pages/common/ProductDestinationPage';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import thunk from 'redux-thunk';
import { ADD_SELECTED_EXPORT_COUNTRY } from '../../../src/client/actions/index';
import i18n from '../../../src/i18n';
import { render } from '@testing-library/react';

global.scrollTo = jest.fn();

describe('Product Destination Page', () => {
  let wrapper;

  const props = {
    history: [],
    route: {
      title: 'Page title',
      header: 'Page header',
      previousUri: ':documentNumber/previous-url',
      nextUri: ':documentNumber/next-url',
      path: ':documentNumber/what-export-destination',
      saveAsDraftUri: '/save-as-draft-url',
      journey: 'journey',
      progressUri: 'document123/progress'
    },
    match: {
      params: {
        documentNumber: 'document123',
      },
    },
  };

  const mockStore = configureStore([
    thunk.withExtraArgument({
      orchestrationApi: {
        get: () => {
          return new Promise((res) => {
            res({});
          });
        },
        post: () => {
          return new Promise((res) => {
            res({});
          });
        },
      },
    }),
  ]);

  let store = mockStore({
    errors: {},
    exportLocation: {
      exportedTo: 'New Zealand',
      loaded : true
    },
    global: {
      allCountries: ['UK', 'Brazil'],
      allCountriesData: [
        {
          officialCountryName: 'United Kingdom',
          isoCodeAlpha2: 'Alpha2',
          isoCodeAlpha3: 'Alpha3',
          isoNumericCode: 'IsoNumericCode'
        },
        {
          officialCountryName: 'Brazil',
          isoCodeAlpha2: 'Alpha2',
          isoCodeAlpha3: 'Alpha3',
          isoNumericCode: 'IsoNumericCode'
        },
        {
          officialCountryName: 'Finland',
          isoCodeAlpha2: 'Alpha2',
          isoCodeAlpha3: 'Alpha3',
          isoNumericCode: 'IsoNumericCode'
        }
      ]
    },
  });

  beforeEach(() => {
    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/what-export-destination']}>
          <ProductDestinationPage {...props} />
        </MemoryRouter>
      </Provider>
    );
  });

  it('will render the correct title', () => {
    expect(wrapper.find('PageTitle').prop('title')).toBe(
      'Page header - Page title'
    );
  });

  it('will render the correct header', () => {
    expect(wrapper.find('h1').text()).toBe(props.route.header);
  });

  it('will render the ExportDestination component', () => {
    expect(wrapper.find('ExportDestination').exists()).toBe(true);
  });

  it('should pass the exportedTo field to the ExportDestination component', () => {
    expect(wrapper.find('ExportDestination').prop('exportDestination')).toBe(
      'New Zealand'
    );
  });

  it('will handle onChange for the ExportDestination component', () => {
    const onChange = wrapper.find('ExportDestination').prop('onChange');

    onChange('Brazil');

    expect(store.getActions()).toContainEqual({
      type: ADD_SELECTED_EXPORT_COUNTRY,
      payload: {
        exportedTo: {
          isoCodeAlpha2: 'Alpha2',
          isoCodeAlpha3: 'Alpha3',
          isoNumericCode: 'IsoNumericCode',
          officialCountryName: 'Brazil',
        },
        loaded: true
      },
    });
  });

  it('will handle onSubmit', () => {
    wrapper.find('button#continue').simulate('submit', { preventDefault() {} });
  });

  it('will not render an error island by default', () => {
    expect(wrapper.find('ErrorIsland').exists()).toBe(false);
  });

  it('renders Back to Your Progress link with the correct href property', () => {
    expect(wrapper.find('BackToProgressLink').find('a')).toBeTruthy();
    expect(wrapper.find('BackToProgressLink').find('a').props().href).toBe('/document123/progress');
  });

  it('should take a snapshot of the whole page', () => {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  });

  describe('if there are errors', () => {
    let wrapperWithErrors;

    const storeWithErrors = mockStore({
      errors: {
        exportDestinationError: i18n.t('commonProductDestinationErrorInvalidCountry'),
        errors: [{targetName:'exportDestination', text:'commonProductDestinationErrorInvalidCountry'}],
      },
      exportLocation: {
        exportedTo: 'New Zealand',
        loaded : true
      },
      global: {
        allCountries: ['UK', 'Brazil'],
      },
    });

    beforeAll(() => {
      wrapperWithErrors = mount(
        <Provider store={storeWithErrors}>
          <MemoryRouter>
            <ProductDestinationPage {...props} />
          </MemoryRouter>
        </Provider>
      );
    });

    it('will render an error island', () => {
      expect(wrapperWithErrors.exists('ErrorIsland')).toBe(true);
    });

    it('will populate the error property on the ExportDestination component', () => {
      expect(wrapperWithErrors.find('ExportDestination').prop('error')).toBe(
        'Select a valid destination country'
      );
    });
  });
});
