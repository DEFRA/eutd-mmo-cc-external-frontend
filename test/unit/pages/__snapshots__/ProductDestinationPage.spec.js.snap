// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Destination Page should take a snapshot of the whole page 1`] = `
<div>
  <div
    class="css-1l99ixu-OuterContainer e4u1u2g0"
  >
    <div
      class="css-n57fe4-InnerContainer e4u1u2g1"
    >
      <a
        class="e1vcssft0 css-zk98nb-Anchor-StyledHoc e4u1u2g0"
        href="undefined/previous-url"
      >
        Back
      </a>
      <h1
        class="e1vcssft0 css-11nzzut-StyledHeader-StyledHoc e4u1u2g0"
      >
        Page header
      </h1>
      <form
        action="/orchestration/api/v1/exportDestination?c=:documentNumber/what-export-destination&n=:documentNumber/next-url"
        method="POST"
      >
        <input
          name="journey"
          type="hidden"
          value="journey"
        />
        <input
          name="currentUri"
          type="hidden"
          value=":documentNumber/what-export-destination"
        />
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <div
            class="css-n58wva-StyledColumn e4u1u2g0"
          >
            <label
              class="e1vcssft0 css-137pbp-StyledLabel-StyledHoc e4u1u2g0"
              for="exportDestination"
            >
              Select the destination country
              <span
                class="e1vcssft0 css-1whl0n4-StyledHint-StyledHoc e4u1u2g0"
              >
                This is the main destination country for the export, not the countries it is passing through. This information will not appear on the final document.
              </span>
              <div
                class="autocomplete__wrapper"
              >
                <div
                  style="border: 0px; height: 1px; margin-bottom: -1px; margin-right: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
                >
                  <div
                    aria-atomic="true"
                    aria-live="polite"
                    id="exportDestination__status--A"
                    role="status"
                  />
                  <div
                    aria-atomic="true"
                    aria-live="polite"
                    id="exportDestination__status--B"
                    role="status"
                  />
                </div>
                
                <input
                  aria-activedescendant="false"
                  aria-autocomplete="list"
                  aria-describedby="exportDestination__assistiveHint"
                  aria-expanded="false"
                  aria-owns="exportDestination__listbox"
                  autocomplete="off"
                  class="autocomplete__input autocomplete__input--default"
                  id="exportDestination"
                  name="exportDestination"
                  placeholder=""
                  role="combobox"
                  type="text"
                  value=""
                />
                <ul
                  class="autocomplete__menu autocomplete__menu--overlay autocomplete__menu--hidden"
                  id="exportDestination__listbox"
                  role="listbox"
                />
                <span
                  id="exportDestination__assistiveHint"
                  style="display: none;"
                >
                  When autocomplete results are available use up and down arrows to review and enter to select.  Touch device users, explore by touch or with swipe gestures.
                </span>
              </div>
            </label>
          </div>
        </div>
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <button
            class="css-17uw1q2-StyledButton-StyledHoc e4u1u2g0"
            formaction="/orchestration/api/v1/exportDestination/saveAsDraftLink"
            id="saveAsDraft"
            name="saveAsDraft"
            type="submit"
          >
            Save as Draft
          </button>
          <button
            class="css-1gc9650-StyledButton-StyledHoc e4u1u2g0"
            id="continue"
            type="submit"
          >
            Save and continue
          </button>
        </div>
      </form>
      <a
        href="/document123/progress"
        id="backToProgress"
      >
        Back to Your Progress
        <span
          class="govuk-visually-hidden"
        >
          back to your progress page link
        </span>
      </a>
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <hr
            class="help-link"
          />
          <h2
            class="e1vcssft0 css-162lx7h-StyledHeader-StyledHoc e4u1u2g0"
          >
            Need help?
          </h2>
          <a
            href="undefined"
            rel="noopener noreferrer"
            target="_blank"
          >
            Get help exporting fish from the UK (gov.uk)
             
            <span
              class="govuk-visually-hidden"
            >
              (opens in new tab)
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
